rfcs/20240326-block-level-api/README.md

```diff
@@ -0,0 +1,638 @@
+  are set, we can create a ukernel object which would have all
+  queriable information.
+- keep a single class for each ukernel functionality, but expose a
+  `finalize()` method after which attributes can no more be changed
+  and queries can be issued.
+
+We propose to go with the latter solution in order to minimize the
+amount of objects created on user side, and minimize overheads.
+
+## Managing jitting overheads and caching support
+
+To implement a single matrix multiplication operation, the block-level
+API user will typically need to rely on multiple kernels.
+In particular:
+- tail handling on M, N and both M/N can require dedicated kernels if
+  M/N are not multiples of the block-size chosen by the user,
+- when blocking on K dimension, extra kernels are required to handle
+  first K-block (beta=0), intermediate K-blocks (beta=1) and last
+  K-block (beta=1, D matrix set, and post-ops specified).
+- when full matrices dimension change frequently. This will lead to
+  changing leading dimension at the brgemm level, requiring new
+  kernels.
+
+On top of that, if the block-level kernels are generated independantly
+for each layer of a neural network, the same kernels might be
+generated multiple times.
+
+Some API design decisions could impact potential reusability of kernels.
+In particular:
+- passing alpha/beta at runtime instead of creation time.
+- support two execute functions (with and without post-op and final
+  data conversion) for the same brgemm object.
+- support runtime leading dimensions.
+
+The proposal here is to stick with what is already tried and used
+internally. If during the experimental stage of the API, users request
+these functionalities, we will revisit the API and add the necessary
+support in the internal implementation. So alpha/beta will be creation
+time parameters, and kernels with and without post-ops will be handled
+with separate objects.  For runtime leading dimension, the internal
+API supports it but this will not be exposed for now. We can either
+change the API when we introduce the feature, or add a placeholder
+parameter if we are confident this feature will gain traction soon.
+
+Also, note that currently, no kernel-level cache exists for block
+level functionalities, but there are plans to add it. In the meantime,
+here are a few ways to mitigate the lack of kernel-level cache:
+- hoist block-level kernel generation for it to happen once at the
+  beggining of the application
+- to handle tails/K-blocking, if generating multiple kernels is too
+  costly, one can copy A/B blocks to a temporary buffer for
+  tails/first K-block, and rely on a single block-level kernel.
+
+
+## Transforms and transpose
+
+Transformation routines, for example to pack data in a hardware required format,
+are typically hard to implement without using intrinsics or assembly.
+To facilitate packing, we will expose an out-of-place transform
+functionality.
+
```

Comment:
Will transform provide the function of doing pack and transpose at the same time?
For matrix B, it not only needs to be packed, but also often needs to be transposed. It would improve performance if it could be done in one pass.

Reply:
Because we take a different pack_type for input and  output, the plan is that we could add `transposed` pack_type.
This way user can set `src_pack = transposed`, and `dst_pack = pack_32` for example.
Does that work for you or do you envision more complex use cases?

Reply:
> Because we take a different pack_type for input and output, the plan is that we could add `transposed` pack_type. This way user can set `src_pack = transposed`, and `dst_pack = pack_32` for example. Does that work for you or do you envision more complex use cases?

Yes. Thanks.